### PR TITLE
Adding support for multi-modal file input for Vertex AI

### PR DESCRIPTION
**Pull request by smwitkowski**
### Context and Motivation

The existing `vertexai_message_parser` function only supports messages with either string content or a list of string content. This limits the flexibility of the Vertex AI client, as it cannot handle messages containing both strings and `gm.Part` objects, which are necessary for including images, videos, documents, and audio files.

This PR introduces support for providing `Part` objects, enhancing the flexibility and functionality of the Vertex AI client.

### Summary of Changes 

- Modified `vertexai_message_parser` to handle messages with mixed content types, including strings and `gm.Part` objects.
- Added unit tests to ensure the correct parsing of messages with mixed content types.

### Detailed Explanation

The updated `vertexai_message_parser` now checks for list content, and if found, iterates through the list items. For each item, it checks if it's a string or a `gm.Part` object. If it's a string, it converts it to a `gm.Part` object. Otherwise, it simply appends the existing `gm.Part` object to the list of parts. This ensures that the final `gm.Content` object contains a list of `gm.Part` objects, correctly representing the message content.

### Testing

New unit tests have been added to `tests/llm/test_vertexai/test_message_parser.py` to cover the following scenarios:

- Parsing a message with string content.
- Parsing a message with a list of string content.
- Parsing a message with a list containing both strings and `gm.Part` objects.
- Handling invalid content types.
- Handling invalid items in a list.

### Relevant Links

- [Issue #946 ](https://github.com/jxnl/instructor/issues/946)
- [Image understanding  |  Generative AI on Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/image-understanding)

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 8de32adcfab67844aa6c50cccdf9c28fcef2656b  | 
|--------|--------|

### Summary:
Enhanced `vertexai_message_parser` to support mixed content types, adding tests for validation.

**Key points**:
- Updated `vertexai_message_parser` in `instructor/client_vertexai.py` to handle mixed content types: strings and `gm.Part` objects.
- Added error handling for unsupported content types in `vertexai_message_parser`.
- Modified `_vertexai_message_list_parser` to support the updated `vertexai_message_parser`.
- Added unit tests in `tests/llm/test_vertexai/test_message_parser.py` for various content scenarios, including mixed content types.
- Added integration tests in `tests/llm/test_vertexai/test_modes.py` to validate mixed content handling in real scenarios.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by smwitkowski** [comment]
> 👍 Looks good to me! Reviewed everything up to [8de32ad](https://github.com/jxnl/instructor/commit/8de32adcfab67844aa6c50cccdf9c28fcef2656b) in 10 seconds
> 
> More details
> * Looked at `142` lines of code in `3` files
> * Skipped `0` files when reviewing.
> * Skipped posting `4` drafted comments based on config settings.
> 
> 1. `instructor/client_vertexai.py:36`
> 2. `instructor/client_vertexai.py:57`
> 3. `instructor/client_vertexai.py:35`
> 4. `tests/llm/test_vertexai/test_message_parser.py:1`
> Workflow ID: `wflow_mRRx9diPpEgJ56lq`
> 
> You can customize Ellipsis with 👍 / 👎 [feedback](https://docs.ellipsis.dev/review), review rules, user-specific overrides, `quiet` mode, and [more](https://docs.ellipsis.dev/config).

Addressed (1) and (2) in https://github.com/jxnl/instructor/pull/947/commits/0c37e7e3cf83dfe71e9f7e79ee659c18d59bf5c8

**Review by AmirNajian — APPROVED** [review]

**Comment by ivanleomk** [comment]
@jxnl Sure thing, will allocate some time to this, just need to understand what the PART is useful for

**Comment by ivanleomk** [comment]
@smwitkowski this is the current status of the tests for me

```bash
tests/llm/test_vertexai/test_message_parser.py ....                                                                             [ 15%]
tests/llm/test_vertexai/test_modes.py ......                                                                                    [ 38%]
tests/llm/test_vertexai/test_retries.py ....                                                                                    [ 53%]
tests/llm/test_vertexai/test_simple_types.py ..FF.F..                                                                           [ 84%]
tests/llm/test_vertexai/test_stream.py FF..                                                                                     [100%]
```

Were you able to get all of the tests passing on your end?

**Comment by smwitkowski** [comment]
@ivanleomk - after rebasing with the changes made in #955 (thanks!) all my tests are passing.

image

Allowing for `Part` types enables users to take advantage of Gemini's multi-modal capabilities. With this change, they can provide documents, images, audio, and video files to Gemini, in addition to plain text.  Google uses `Part` to facilitate the transport of multi-modal content. 

https://cloud.google.com/vertex-ai/generative-ai/docs/reference/python/latest/vertexai.generative_models.Part

**Comment by ivanleomk** [comment]
Awesome! I spent yesterday fixing some bugs and now it should work. I tested it with the following code snippets

```python
import instructor
import vertexai.generative_models as gm  # type: ignore
from pydantic import BaseModel, Field
import requests

client = instructor.from_vertexai(gm.GenerativeModel("gemini-1.5-pro-001"))
content = [
    "Order Details:",
    gm.Part.from_text("Customer: Alice"),
    gm.Part.from_text("Items:"),
    "Name: Laptop, Price: 999.99",
    "Name: Mouse, Price: 29.99",
]


class Item(BaseModel):
    name: str
    price: float


class Order(BaseModel):
    items: list[Item] = Field(..., default_factory=list)
    customer: str


resp = client.create(
    response_model=Order,
    messages=[
        {
            "role": "user",
            "content": content,
        },
    ],
)
```

```
client = instructor.from_vertexai(
    gm.GenerativeModel("gemini-1.5-pro-001"), mode=instructor.Mode.VERTEXAI_JSON
)
content = [
    gm.Part.from_text("Count the number of objects in the image."),
    gm.Part.from_data(
        bytes(
            requests.get(
                "https://img.taste.com.au/Oq97xT-Q/taste/2016/11/blueberry-scones-75492-1.jpeg"
            ).content
        ),
        "image/jpeg",
    ),
]


class Description(BaseModel):
    description: str


resp = client.create(
    response_model=Description,
    messages=content,
)

print(resp)
```

These both seem to work nicely. @smwitkowski I think this looks good to merge. Could I trouble you to add one more test for image understanding ( just make sure to use the JSON mode, function calling doesn't support it for now ) and we can merge it in?

File: tests/llm/test_vertexai/test_modes.py [hunk 1]
```diff
@@ -92,3 +92,30 @@ def test_complex_nested_model(model, mode):
     assert {book.author for book in resp.books} == {"Jane Doe", "John Smith"}
     assert {book.genre for book in resp.books} == {"Fantasy", "Non-Fiction"}
     assert {book.isbn for book in resp.books} == {"1234567890", "0987654321"}
+
```
Comment (ivanleomk) [inline tests/llm/test_vertexai/test_modes.py:95]: Let's add one more test for image understanding and should be good to merge. Going to just approve it first

**Review by ivanleomk — APPROVED** [review]

File: tests/llm/test_vertexai/test_modes.py (hunk 1 above)
Comment (ivanleomk) [reply tests/llm/test_vertexai/test_modes.py:95]: Maybe something like this just to make sure we don't throw any errors in the `test_vertexai` folder

```
client = instructor.from_vertexai(
    gm.GenerativeModel("gemini-1.5-pro-001"), mode=instructor.Mode.VERTEXAI_JSON
)
content = [
    gm.Part.from_text("What country is this in?"),
    gm.Part.from_uri(
        "gs://cloud-samples-data/vertex-ai/llm/prompts/landmark2.png",
        mime_type="image/png",
    ),
]


class Description(BaseModel):
    description: str


resp = client.create(
    response_model=Description,
    messages=content,
)

print(resp)
```

**Pull request merged**